### DISCUSSION TITLE
Official/unofficial parts load/search order

### DISCUSSION THREAD

**Discussion by trevorsandy** [leozide/leocad#652] [Ideas]
When the UNOFFICIAL library is present, loaded parts will ultimately point to this archive/directory if those parts are present in the OFFICIAL library. The offending code is [here](https://github.com/leozide/leocad/blob/57c4f2912dfedec9df27cad909765f40b0809ff1/common/lc_library.cpp#L451) for archive libraries, and [here](https://github.com/leozide/leocad/blob/57c4f2912dfedec9df27cad909765f40b0809ff1/common/lc_library.cpp#L525) and [here](https://github.com/leozide/leocad/blob/57c4f2912dfedec9df27cad909765f40b0809ff1/common/lc_library.cpp#L700) for disc libraries. This is not usually an expected behaviour because the maturity of unofficial parts may be less than their official counterpart and the behaviour (e.g. orientation) may also be different.

Although this scenario (official and unofficial files with the same name and different behaviour) is an uncommon exception and should actually not be permitted by the LDraw administrators (in their released libraries), it is a valid use case, as we can see [here](https://github.com/trevorsandy/lpub3d/issues/508), so I’ll add a flag in my LeoCAD library for LPub3D preference dialogue to allow the user to specify the official/unofficial search order when the unofficial library is present. Other popular LDraw editors and viewers either default to the official part or offer a search order choice.

The thinking behind adding a search order preference flag versus simply defaulting to the official part takes into account the scenario where the user may intentionally have modified official parts in his/her unofficial library. Moreover, looking at the code, loading/searching the unofficial libraries first when they are present is clearly intentional.

If you believe this update is useful, let me know and I’ll author a PR for LeoCAD.

Cheers,

**Comment by rsbx** [comment]
Interesting. I'd expected (I only have the latest OFFICIAL parts atm) 
the the search order to OFFICIAL then UNOFFICIAL to fill in.

**Comment by leozide** [comment]
The order was intentional, I thought that unofficial parts could have fixes for official parts so they should have priority. I've tried to get the official answer from LDraw but I couldn't: https://forums.ldraw.org/thread-23765-post-34726.html

I think it's ok to change the order for everyone, I don't think need an option for this.

**Comment by rsbx** [comment]
That forum exchange was pretty un-enlightening.

Another option is to have a settable search path.

**Comment by trevorsandy** [comment]
> I think it's ok to change the order for everyone...

Very well. This update is quite trivial so just I'll close the submitted PR.

Cheers,

**Comment by leozide** [comment]
I'm tracking this today, looks like LDView searches official parts first too so it either changed or the post is incorrect. https://github.com/tcobbs/ldview/blob/6407a79c2f3a77710afa4cd8b24bb298bce3c76b/LDLoader/LDrawIni.c#L430

**Comment by trevorsandy** [comment]
> I'm tracking this today, looks like LDView searches official parts first too so it either changed or the post is incorrect.

I don’t understand. What has changed or is incorrectly posted?

Indeed, my understanding LDView by default will load/lookup official parts and does not lookup unofficial parts unless the unofficial directory is explicitly specified. The code at the link you provided is triggered when LDraw.ini is present but search directory entries are not specified. Nevertheless, it confirms the default load/lookup order starts with official parts.

For Windows and MacOS, you can specify extra search directories from the extra search directories menu action and LDView.ini file. On Unix, it can be done via the LDView.ini file.

Cheers,

**Reply by leozide** [reply]
> I don’t understand. What has changed or is incorrectly posted?

The conversation in the LDraw forums had unofficial folders first.

**Reply by trevorsandy** [reply]
Ah - got it. 

I believe both your original premise  “_I thought that unofficial parts could have fixes for official parts so they should have priority_” and mine “_the maturity of unofficial parts may be less than their official counterpart and the behaviour (e.g. orientation) may also be different_” are quite valid so an option to choose/set the order is better than a single fixed behaviour.

Consequently, I’ve completed the official/unofficial load order update option with commits [50d32cd](https://github.com/trevorsandy/leocad/commit/50d32cd4084b851eece283b0cd021aa664b7a1db) and [46d7da4](https://github.com/trevorsandy/leocad/commit/46d7da4ac3aaa49e48aa6506eb876399b234dc2b). These updates are not reflected in the closed PR.

Cheers,

**Comment by rsbx** [comment]
Code talks...
